pkg/reconciler/nodeconfig_reconciler.go

```diff
@@ -0,0 +1,413 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: node.Name,
+		},
+	}
+
+	c.Spec = *revision.Spec.Config.DeepCopy()
+	c.Spec.Revision = revision.Spec.Revision
+	c.Name = node.Name
+
+	if err := controllerutil.SetOwnerReference(node, c, scheme.Scheme); err != nil {
+		return nil, fmt.Errorf("error setting owner references (node): %w", err)
+	}
+
+	if err := controllerutil.SetOwnerReference(revision, c, ncr.scheme); err != nil {
+		return nil, fmt.Errorf("error setting owner references (revision): %w", err)
+	}
+
+	// prepare Layer2NetworkConfigurationSpec (l2Spec) for each node.
+	// Each Layer2NetworkConfigurationSpec from l2Spec has node selector,
+	// which should be used to add config to proper nodes.
+	// Each Layer2NetworkConfigurationSpec that don't match the node selector
+	// is removed.
+	for i := 0; i < len(c.Spec.Layer2); i++ {
+		if c.Spec.Layer2[i].NodeSelector == nil {
+			// node selector is not defined for the spec.
+			// Layer2 is global - just continue
+			continue
+		}
+
+		// node selector of type v1.labelSelector has to be converted
+		// to labels.Selector type to be used with controller-runtime client
+		selector, err := convertSelector(c.Spec.Layer2[i].NodeSelector.MatchLabels, c.Spec.Layer2[i].NodeSelector.MatchExpressions)
+		if err != nil {
+			return nil, fmt.Errorf("error converting selector: %w", err)
+		}
+
+		// remove currently processed Layer2NetworkConfigurationSpec if node does not match the selector
+		if !selector.Matches(labels.Set(node.ObjectMeta.Labels)) {
+			// TODO: is it worth to preserve order?
+			c.Spec.Layer2 = append(c.Spec.Layer2[:i], c.Spec.Layer2[i+1:]...)
+			i--
+		}
+	}
+
+	// set config as next config for the node
+	return c, nil
+}
+
+func convertSelector(matchLabels map[string]string, matchExpressions []metav1.LabelSelectorRequirement) (labels.Selector, error) {
+	selector := labels.NewSelector()
+	var reqs labels.Requirements
+
+	for key, value := range matchLabels {
+		requirement, err := labels.NewRequirement(key, selection.Equals, []string{value})
+		if err != nil {
+			return nil, fmt.Errorf("error creating MatchLabel requirement: %w", err)
+		}
+		reqs = append(reqs, *requirement)
+	}
+
+	for _, req := range matchExpressions {
+		lowercaseOperator := selection.Operator(strings.ToLower(string(req.Operator)))
+		requirement, err := labels.NewRequirement(req.Key, lowercaseOperator, req.Values)
+		if err != nil {
+			return nil, fmt.Errorf("error creating MatchExpression requirement: %w", err)
+		}
+		reqs = append(reqs, *requirement)
+	}
+	selector = selector.Add(reqs...)
+
+	return selector, nil
+}
+
+func (ncr *NodeConfigReconciler) deployConfig(ctx context.Context, newConfig, currentConfig *v1alpha1.NodeNetworkConfig, node *corev1.Node) error {
+	var cfg *v1alpha1.NodeNetworkConfig
+	if currentConfig != nil {
+		cfg = currentConfig
+		// there already is config for node - update
+		cfg.Spec = *newConfig.Spec.DeepCopy()
+		cfg.Status = *newConfig.Status.DeepCopy()
```

Comment:
do we need to deep copy here? Also, status should probably be cleared instead of copied here. Updating it is the task of the agent.

Reply:
You're right, there's no need for a deep copy. 
Status will only be cleared here and left to agent for further processing.